# App Store Analysis

## Using Data Analytics to Inform Business Strategy

## 1. Project Design

### 1.1 Company Background

The company XYZ is in the business of building Android and iOS mobile apps. These apps are free to download and install, and they are distributed through Google Play Store and the Apple App Store.

The company's main source of revenue is in-app advertising. This posits the business model as volume driven - scale in terms of the number of users becomes very important. The more the number of users who see and engage with the ads, proportionally better is the revenue opportunity.

### 1.2 Business Challenge

The senior management team is meeting for the annual strategy event to decide on allocation of resources and future app development roadmap. The team is seeking inputs from the business strategy group that will help the company maximize return-on-investment(ROI) opportunities. 

### 1.3 Project Scope

Our goal for this project is to offer actionable insights that are backed by data. Based on our understanding of the company's business model, we know that the biggest driver of ROI is the number of users for an app - the revenue opportunity is directly proportional. We will focus our exploration on this topic. 

Our project scope is to analyze app store data and identify the type of apps that are likely to attract more users. Such actionable intelligence can help optimize revenue and the company can focus on creating the kind of apps that are popular.

Our key requirements are as follows:

- We are interested in free apps only
- We are interested in apps in English language only


### 1.4 Sources of data

Apple Store Data: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps. This dataset contains data about ~7000 iOS apps as of July 2017.

Google Play Store Data: https://www.kaggle.com/lava18/google-play-store-apps. This dataset contains data about approximately 10,000 Android apps as of August 2018.

## 2. Data Preparation

### 2.1 Open Apple Play Store and Google Play Store data sets

In [1]:
# Open both datasets and save them as list of lists.

from csv import reader

#Apple Store Dataset
opened_file = open("./AppleStore.csv")
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

#Google Play Store Dataset
opened_file = open("./googleplaystore.csv")
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

### 2.2 Create explore_data() function

In [2]:
# Defining a function to make it easy to print data

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    
    for row in dataset_slice:
        print(row)
        print("\n") #Add a new empty line after each row
        
    if rows_and_columns:
        print("Number of rows: " + str(len(dataset)))
        print("Number of columns: " + str(len(dataset[0])))

### 2.3 Exploring the datasets

Let's look at the structure of the two datasets that we have created. For each dataset, we would like to know the following:

- Number of columns in each dataset to learn about the headers
- Number of rows in each dataset to learn about the total number of entries

Let's use the explore_data() function that we created to gather these insights.

#### 2.3.1 Exploring the iOS dataset

In [3]:
print("ios header:\n", ios_header, "\n")
explore_data(ios, 0, 2, True)

ios header:
 ['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7197
Number of columns: 16


#### 2.3.2 Exploring the android dataset

In [4]:
print("android header:\n", android_header, "\n")
explore_data(android, 0, 2, True)

android header:
 ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


#### 2.3.3 Results of the data structure analysis

| Data set | Number of Rows | Numer of Columns | Column Names |
| ------   | ------         | ------       | ------ |
| iOS      | 7197           | 16 | 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'     |
| Android  | 10841          | 13 | 'App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'|

## 3. Data Cleansing

Our two datasets, in their current format, are a list of lists. However, we cannot use them right away. The data needs to be cleaned and prepared so that we do not get any wrong results in our analysis. As per our requirements, we need to remove all paid apps and non-English language apps too.

We will focus on the following three steps that are integral to any data cleaning process:

- remove or correct wrong data
- remove duplicate data
- modify the data to fit the purpose of our analysis

### 3.1 Finding and deleting erroneous data in Google Play Store dataset

#### 3.1.1 Deleting wrong data

In the [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion) of the Google Play Store dataset, [this discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) on Google Play Store describes missing data for row 10472. Let's check if this is the case by matching the length of the entry 10472 to the length of the header.

In [5]:
# Checking the index of the entry with missing data

for row in android:
    if len(row) != len(android_header):
        print("Row with wrong data: ", android.index(row))
        print(row)

Row with wrong data:  10472
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Let us find what's wrong with the data in this row. We will compare this row with another row from the dataset to find the anomaly.

In [6]:
print("android header: :\n", android_header, "\n")
print("Row #1 of the data set: \n", android[0], "\n")

android header: :
 ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

Row #1 of the data set: 
 ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 



As we compare rows 1 and 10472, we find that row 10472 has a missing Category, and all other data has moved left by one column (for example, the reviews column for row 10472 shows '3.0M' which is wrong). We need to remove this row as the wrong data will create errors in our analysis.

In [7]:
print("Total number of rows before deletion: ", len(android))
del android[10472] #do not run this more than once
print("Total number of rows after deletion: ", len(android))

Total number of rows before deletion:  10841
Total number of rows after deletion:  10840


#### 3.1.2 Deleting duplicate entries

Let's find if there are any duplicate entries in the dataset. We can do this by creating two lists, viz., one list containing unique apps and the other list containing duplicate apps. Further to this, we will dig deeper into the duplicate apps list to find a way to select the right data in the android dataset and remove duplicate entries.

In [8]:
android_clean = []    #create an empty list to store unique apps
android_duplicate = []    #create an empty list to store duplicate apps

for row in android:
    if row[0] in android_clean:    #checking by App Name
        android_duplicate.append(row[0])
    else:
        android_clean.append(row[0])
        
print("Number of rows in the cleaned dataset: ", len(android_clean))
print("Number of rows in the duplicate dataset: ", len(android_duplicate))

duplicate_counts = {}

for app in android_duplicate:
    if app in duplicate_counts:
        duplicate_counts[app] += 1
    else:
        duplicate_counts[app] = 1

print("App with the highest number of duplicate entries:", max(duplicate_counts, key=duplicate_counts.get))

Number of rows in the cleaned dataset:  9659
Number of rows in the duplicate dataset:  1181
App with the highest number of duplicate entries: ROBLOX


Our initial analysis shows that there are 1181 duplicate entries in our dataset. We have also found that the ROBLOX app has the highest number of duplicate entries. Let's print all entries with the ROBLOX app name in the android dataset to identify which entry is most relevant and also find a way to remove duplicates.

In [9]:
for app in android:
    if app[0] == "ROBLOX":
        print(app)

['ROBLOX', 'GAME', '4.5', '4447388', '67M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Adventure;Action & Adventure', 'July 31, 2018', '2.347.225742', '4.1 and up']
['ROBLOX', 'GAME', '4.5', '4447346', '67M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Adventure;Action & Adventure', 'July 31, 2018', '2.347.225742', '4.1 and up']
['ROBLOX', 'GAME', '4.5', '4448791', '67M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Adventure;Action & Adventure', 'July 31, 2018', '2.347.225742', '4.1 and up']
['ROBLOX', 'GAME', '4.5', '4449882', '67M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Adventure;Action & Adventure', 'July 31, 2018', '2.347.225742', '4.1 and up']
['ROBLOX', 'GAME', '4.5', '4449910', '67M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Adventure;Action & Adventure', 'July 31, 2018', '2.347.225742', '4.1 and up']
['ROBLOX', 'FAMILY', '4.5', '4449910', '67M', '100,000,000+', 'Free', '0', 'Everyone 10+', 'Adventure;Action & Adventure', 'July 31, 2018', '2.347.225742', '4.1

We notice that the entries differ based on the total number of reviews, which is column 4 (index 3). We can form a hypothesis to sort duplicates by reviews - the higher the number of reviews, the recent the data should be.

In [10]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print("Expected length of clean list: ", len(android) - len(android_duplicate))
print("Actual length of the new sorted data: ", len(reviews_max))

Expected length of clean list:  9659
Actual length of the new sorted data:  9659


In [11]:
android_clean = []    #stores the new cleaned dataset
already_added = []    #stores app names to avoid duplicates in case the review count is same

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

In [12]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


### 3.2 Deleting wrong data in Apple Store dataset

[This discussion](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion/106176) on App Store data mentions presence of duplicate data. Let's find the duplicate data in our ios dataset.

#### 3.2.1 Deleting wrong data and duplicates

In [13]:
print(ios_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [14]:
# Create two empty lists - one list contains unique data and the other list contains duplicates.

ios_clean = []
ios_duplicate_apps = []

print("Number of rows in the original dataset: ", len(ios))
for app in ios:
    if app[0] in ios_clean:    #checking by id column
        ios_duplicate_apps.append(app[0])
    else:
        ios_clean.append(app[0])
        
print("Number of rows in the cleaned dataset: ", len(ios_clean))
print("Number of rows in the duplicate dataset: ", len(ios_duplicate_apps))

Number of rows in the original dataset:  7197
Number of rows in the cleaned dataset:  7197
Number of rows in the duplicate dataset:  0


The App Store dataset does not seem to have any app with duplicate entries.

### 3.3 Removing non-English Apps

As mentioned in Section 1.3, one of the key requirements of our project is to focus on apps that are in English language only. As we analyze our datasets, we find that are many apps that are not designed for English speaking audiences. We will remove these apps from the list by first identifying them and then removing them.